File: bench/line.go
// SQZ v2 line grammar — Go mirror of src/line.ts (tokenizer + validator).
// The wire format is ONE plain line: <mode> Δ["f1","f2"] [L:<lang>]
// <symbol>[operand]... v"verbatim clause"... — no JSON envelope.

package bench

import "strings"

// LineClause is one symbol clause in a SQZ line.
type LineClause struct {
	Symbol  string
	Operand string
}

// LineAST is the parsed form of a SQZ line.
type LineAST struct {
	Mode     string
	Files    []string
	Lang     string
	Clauses  []LineClause
	Verbatim []string
	Errors   []string
}

var lineGlyphs = []rune{'Δ', '≋', '∂', 'μ', '⌁', '→', '✓', '⏭'}

func isLineGlyphRune(r rune) bool {
	for _, g := range lineGlyphs {
		if r == g {
			return true
		}
	}
	return false
}

// readLineQuoted reads a "..." quoted string starting at rs[i] (the quote).
func readLineQuoted(rs []rune, i int, errors *[]string) (string, int) {
	var out strings.Builder
	j := i + 1
	for j < len(rs) {
		ch := rs[j]
		if ch == '\\' && j+1 < len(rs) {
			out.WriteRune(rs[j+1])
			j += 2
			continue
		}
		if ch == '"' {
			return out.String(), j + 1
		}
		out.WriteRune(ch)
		j++
	}
	*errors = append(*errors, "unterminated quote")
	return out.String(), len(rs)
}

// readLineGroup reads a [..] or (..) balanced group starting at rs[i].
func readLineGroup(rs []rune, i int, errors *[]string) (string, int) {
	open := rs[i]
	close := rune(']')
	if open == '(' {
		close = ')'
	}
	depth := 0
	var inner strings.Builder
	for j := i; j < len(rs); j++ {
		ch := rs[j]
		if ch == '"' {
			q, next := readLineQuoted(rs, j, errors)
			inner.WriteRune('"')
			inner.WriteString(q)
			inner.WriteRune('"')
			j = next - 1
			continue
		}
		if ch == open {
			depth++
			if depth > 1 {
				inner.WriteRune(ch)
			}
			continue
		}
		if ch == close {
			depth--
			if depth == 0 {
				return inner.String(), j + 1
			}
			inner.WriteRune(ch)
			continue
		}
		inner.WriteRune(ch)
	}
	*errors = append(*errors, "unbalanced "+string(open))
	return inner.String(), len(rs)
}

// splitLineFiles splits a Δ[...] group inner into file strings.
func splitLineFiles(inner string) []string {
	rs := []rune(inner)
	var parts []string
	var cur strings.Builder
	i := 0
	for i < len(rs) {
		ch := rs[i]
		if ch == '"' {
			var errs []string
			q, next := readLineQuoted(rs, i, &errs)
			parts = append(parts, q)
			i = next
			continue
		}
		if ch == ',' {
			if strings.TrimSpace(cur.String()) != "" {
				parts = append(parts, strings.TrimSpace(cur.String()))
			}
			cur.Reset()
			i++
			continue
		}
		cur.WriteRune(ch)
		i++
	}
	if strings.TrimSpace(cur.String()) != "" {
		parts = append(parts, strings.TrimSpace(cur.String()))
	}
	return parts
}

func isSpaceRune(r rune) bool { return r == ' ' || r == '\t' }
func isStopRune(r rune) bool {
	return isSpaceRune(r) || r == '"' || r == '[' || r == ']' || r == '(' || r == ')'
}

// ParseLine tokenizes a v2 SQZ line into an AST. Never panics, never loops.
func ParseLine(line string, knownSymbols map[string]bool) LineAST {
	ast := LineAST{}
	rs := []rune(strings.TrimSpace(line))
	i := 0
	for i < len(rs) {
		ch := rs[i]
		if isSpaceRune(ch) {
			i++
			continue
		}

		// Verbatim segment: v"raw prose"
		if ch == 'v' && i+1 < len(rs) && rs[i+1] == '"' {
			q, next := readLineQuoted(rs, i+1, &ast.Errors)
			ast.Verbatim = append(ast.Verbatim, q)
			i = next
			continue
		}

		// Stray quoted string — flag it.
		if ch == '"' {
			q, next := readLineQuoted(rs, i, &ast.Errors)
			ast.Errors = append(ast.Errors, `stray quoted token "`+q+`"`)
			i = next
			continue
		}

		// Glyph (possibly with bracket/paren operand).
		if isLineGlyphRune(ch) {
			j := i + 1
			for j < len(rs) && !isStopRune(rs[j]) {
				j++
			}
			symbol := string(rs[i:j])
			if j < len(rs) && (rs[j] == '[' || rs[j] == '(') {
				inner, next := readLineGroup(rs, j, &ast.Errors)
				if symbol == "Δ" && len(ast.Files) == 0 {
					ast.Files = splitLineFiles(inner)
					if len(ast.Files) == 0 {
						ast.Errors = append(ast.Errors, "Δ target has no files")
					}
				} else {
					ast.Clauses = append(ast.Clauses, LineClause{Symbol: symbol, Operand: strings.TrimSpace(inner)})
				}
				i = next
				continue
			}
			if !knownSymbols[symbol] {
				ast.Errors = append(ast.Errors, "unknown symbol "+symbol)
			}
			ast.Clauses = append(ast.Clauses, LineClause{Symbol: symbol})
			i = j
			continue
		}

		// Lang token: L:<token>
		if ch == 'L' && i+1 < len(rs) && rs[i+1] == ':' {
			j := i + 2
			for j < len(rs) && !isSpaceRune(rs[j]) {
				j++
			}
			lang := string(rs[i+2 : j])
			if lang != "" {
				ast.Lang = lang
			} else {
				ast.Errors = append(ast.Errors, "L: has no lang token")
			}
			i = j
			continue
		}

		// Bare word — first one must be a valid mode.
		j := i
		for j < len(rs) && !isStopRune(rs[j]) {
			j++
		}
		if j == i {
			// Stray structural char ([ ] ( )): skip it, flag it — never loop.
			ast.Errors = append(ast.Errors, `unexpected token "`+string(rs[i])+`"`)
			i++
			continue
		}
		word := string(rs[i:j])
		if ast.Mode == "" && modes[word] {
			ast.Mode = word
			i = j
			continue
		}
		ast.Errors = append(ast.Errors, `unexpected token "`+word+`"`)
		i = j
	}
	if ast.Mode == "" {
		ast.Errors = append(ast.Errors, "missing mode")
	}
	if len(ast.Files) == 0 {
		ast.Errors = append(ast.Errors, "missing Δ target")
	}
	return ast
}

// ValidateLine checks a v2 SQZ line against the lexicon symbol set.
func ValidateLine(line string, knownSymbols map[string]bool) (bool, []string) {
	ast := ParseLine(line, knownSymbols)
	if len(ast.Errors) > 0 {
		return false, ast.Errors
	}
	return true, nil
}

// lineKnownSymbols builds the lexicon symbol whitelist used for validation.
func lineKnownSymbols() map[string]bool {
	m := make(map[string]bool, len(lineGlyphs))
	for _, g := range lineGlyphs {
		m[string(g)] = true
	}
	return m
}

File: src/line.ts
/**
 * SQZ v2 line grammar — tokenizer + AST.
 *
 * Wire format is ONE plain line (no JSON envelope):
 *
 *   <mode> Δ["f1","f2"] [L:<lang>] <symbol>[op]... v"verbatim clause"...
 *
 *   mode    := refactor | api | debug | docs | test | review | arch | general
 *   target  := Δ["file"(,"file")*]          (required, first Δ = target)
 *   lang    := L:<token>                    (optional)
 *   clause  := symbol [operand]             operand = [..] or (..) group
 *   verbatim:= v"raw prose"                 (lossless fallback, never dropped)
 *
 * The tokenizer never throws: parse problems are surfaced as `errors` so the
 * translator ladder can retry or pass the original prose through untouched.
 */

import type { Lexicon, Mode } from "./types.js";
import { MODES } from "./types.js";

export interface LineClause {
  symbol: string;
  operand: string;
}

export interface LineAST {
  mode: string;
  files: string[];
  lang: string;
  clauses: LineClause[];
  verbatim: string[];
  errors: string[];
}

/** Symbols with built-in expansion templates (mirror of translator CLAUSE_TEMPLATES). */
export const LINE_GLYPHS = ["Δ", "≋", "∂", "μ", "⌁", "→", "✓", "⏭"] as const;

function isGlyph(ch: string): boolean {
  return (LINE_GLYPHS as readonly string[]).includes(ch);
}

/** Read a "..." quoted string starting at s[i]. Returns [value, nextIndex]. */
function readQuoted(s: string, i: number, errors: string[]): [string, number] {
  // i is at the opening quote
  let out = "";
  let j = i + 1;
  while (j < s.length) {
    const ch = s[j];
    if (ch === "\\" && j + 1 < s.length) {
      out += s[j + 1];
      j += 2;
      continue;
    }
    if (ch === '"') return [out, j + 1];
    out += ch;
    j++;
  }
  errors.push("unterminated quote");
  return [out, s.length];
}

/** Read a [..] or (..) balanced group starting at s[i]. Returns [inner, nextIndex]. */
function readGroup(s: string, i: number, errors: string[]): [string, number] {
  const open = s[i];
  const close = open === "[" ? "]" : ")";
  let depth = 0;
  let inner = "";
  for (let j = i; j < s.length; j++) {
    const ch = s[j];
    if (ch === '"') {
      const [q, next] = readQuoted(s, j, errors);
      inner += '"' + q + '"';
      j = next - 1;
      continue;
    }
    if (ch === open) {
      depth++;
      if (depth > 1) inner += ch; // nested group: keep it
      continue;
    }
    if (ch === close) {
      depth--;
      if (depth === 0) return [inner, j + 1];
      inner += ch; // nested close: keep it
      continue;
    }
    inner += ch;
  }
  errors.push(`unbalanced ${open}`);
  return [inner, s.length];
}

/** Split a Δ[...] group inner into file strings. Handles quoted + bare entries. */
export function splitFiles(inner: string): string[] {
  const parts: string[] = [];
  let cur = "";
  let i = 0;
  while (i < inner.length) {
    const ch = inner[i];
    if (ch === '"') {
      const errors: string[] = [];
      const [q, next] = readQuoted(inner, i, errors);
      parts.push(q);
      i = next;
      continue;
    }
    if (ch === ",") {
      if (cur.trim()) parts.push(cur.trim());
      cur = "";
      i++;
      continue;
    }
    cur += ch;
    i++;
  }
  if (cur.trim()) parts.push(cur.trim());
  return parts.filter((p) => p.length > 0);
}

/** Tokenize a v2 SQZ line into an AST. Never throws. */
export function parseLine(line: string, lexicon: Lexicon): LineAST {
  const ast: LineAST = {
    mode: "",
    files: [],
    lang: "",
    clauses: [],
    verbatim: [],
    errors: [],
  };
  const knownSymbols = new Set(lexicon.map((e) => e.symbol));

  const s = line.trim();
  let i = 0;
  while (i < s.length) {
    const ch = s[i];
    if (/\s/.test(ch)) {
      i++;
      continue;
    }

    // Verbatim segment: v"raw prose"
    if (ch === "v" && s[i + 1] === '"') {
      const [q, next] = readQuoted(s, i + 1, ast.errors);
      ast.verbatim.push(q);
      i = next;
      continue;
    }

    // Quoted string without symbol (stray) — keep literal, flag.
    if (ch === '"') {
      const [q, next] = readQuoted(s, i, ast.errors);
      ast.errors.push(`stray quoted token "${q}"`);
      i = next;
      continue;
    }

    // Δ[...] target (first Δ) or clause with bracket/paren operand.
    if (isGlyph(ch)) {
      let j = i + 1;
      while (j < s.length && !/\s|["[\]()]/.test(s[j])) j++;
      const symbol = s.slice(i, j);
      if (s[j] === "[" || s[j] === "(") {
        const [inner, next] = readGroup(s, j, ast.errors);
        if (symbol === "Δ" && ast.files.length === 0) {
          ast.files = splitFiles(inner);
          if (ast.files.length === 0) ast.errors.push("Δ target has no files");
        } else {
          ast.clauses.push({ symbol, operand: inner.trim() });
        }
        i = next;
        continue;
      }
      if (!knownSymbols.has(symbol)) {
        ast.errors.push(`unknown symbol ${symbol}`);
        ast.clauses.push({ symbol, operand: "" });
      } else {
        ast.clauses.push({ symbol, operand: "" });
      }
      i = j;
      continue;
    }

    // Lang token: L:<token>
    if (ch === "L" && s[i + 1] === ":") {
      let j = i + 2;
      while (j < s.length && !/\s/.test(s[j])) j++;
      const lang = s.slice(i + 2, j);
      if (lang) ast.lang = lang;
      else ast.errors.push("L: has no lang token");
      i = j;
      continue;
    }

    // Mode word (bare) — first bare word must be a valid mode.
    let j = i;
    while (j < s.length && !/\s|["[\]()]/.test(s[j])) j++;
    if (j === i) {
      // Stray structural char ([ ] ( )): skip it, flag it — never loop.
      ast.errors.push(`unexpected token "${s[i]}"`);
      i++;
      continue;
    }
    const word = s.slice(i, j);
    if (!ast.mode && (MODES as readonly string[]).includes(word)) {
      ast.mode = word as Mode;
      i = j;
      continue;
    }
    ast.errors.push(`unexpected token "${word}"`);
    i = j;
  }

  if (!ast.mode) ast.errors.push("missing mode");
  if (ast.files.length === 0) ast.errors.push("missing Δ target");
  return ast;
}

/** Lexicon lookup by symbol (undefined when the symbol is unknown). */
export function lexiconEntry(lexicon: Lexicon, symbol: string) {
  return lexicon.find((e) => e.symbol === symbol);
}
